### PR TITLE
[ISSUE #4158] add request fail reason

### PR DESCRIPTION
**Pull request by yuz10**
**Make sure set the target branch to `develop`**

## What is the purpose of the change

XXXXX

## Brief changelog

#4158

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

File: remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingAbstract.java [hunk 1]
```diff
@@ -496,11 +497,12 @@ public void operationComplete(ChannelFuture f) throws Exception {
         }
     }
 
-    private void requestFail(final int opaque) {
+    private void requestFail(final int opaque, Throwable cause) {
         ResponseFuture responseFuture = responseTable.remove(opaque);
         if (responseFuture != null) {
             responseFuture.setSendRequestOK(false);
             responseFuture.putResponse(null);
+            responseFuture.setCause(cause);
```
Comment (lwclover) [inline remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingAbstract.java:505]: I think we should add a nonnull judgment.

File: remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingAbstract.java [hunk 2]
```diff
@@ -522,7 +524,7 @@ protected void failFast(final Channel channel) {
             if (entry.getValue().getProcessChannel() == channel) {
                 Integer opaque = entry.getKey();
                 if (opaque != null) {
-                    requestFail(opaque);
+                    requestFail(opaque, new RemotingException("channel closed"));
```
Comment (lwclover) [inline remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingAbstract.java:527]: 'failFast' method may be a right call, the second param is null

**Comment by lwclover** [comment]
I think the reason may be that the channel was closed before the pull message request was sent.

**Comment by yuz10** [comment]
> I think the reason may be that the channel was closed before the pull message request was sent.

Yes, the exception occurs after shutdown a consumer.

**Pull request closed without merging**